collector/accelerators.go

```diff
@@ -0,0 +1,140 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type cardData struct {
+	vendor string
+	model  string
+}
+
+type vendorData struct {
+	vendorName string
+	devicesIDs map[string]string
+}
+
+type acceleratorsCollector struct {
+	pciDevicesPath string
+	logger         log.Logger
+}
+
+func init() {
+	registerCollector("accelerator", defaultEnabled, NewAcceleratorCollector)
+}
+
+// NewAcceleratorCollector returns a new Collector exposing accelerator cards count.
+func NewAcceleratorCollector(logger log.Logger) (Collector, error) {
+	return &acceleratorsCollector{
+		pciDevicesPath: filepath.Join(*sysPath, "bus/pci/devices"),
+		logger:         logger,
+	}, nil
+}
+
+var (
+	acceleratorCardsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "accelerator", "cards"),
+		"Number of accelerator cards per vendor and model",
+		[]string{"vendor", "model"}, nil,
+	)
+
+	nvidiaDeviceIDsMap = map[string]string{
+		"0x20b5": "A100",
+		"0x2230": "RTX_A6000",
+		"0x2717": "RTX_4090",
+		"0x2235": "A40",
+		"0x1df5": "V100",
+	}
+
+	// vendor map, add any new vendor to this map
+	vendorToDeviceMap = map[string]vendorData{
+		// nvidia devices
+		"0x10de": vendorData{"NVIDIA", nvidiaDeviceIDsMap},
+	}
+)
+
+func (a *acceleratorsCollector) Update(ch chan<- prometheus.Metric) error {
+	devicesCount := map[cardData]int{}
+	pciDevices, err := os.ReadDir(a.pciDevicesPath)
+	if err != nil {
+		return fmt.Errorf("failed to read from  %q: %w", a.pciDevicesPath, err)
+	}
+
+	for _, pciDevice := range pciDevices {
+		vendorID, err := a.getVendorID(pciDevice)
+		if err != nil {
+			level.Error(a.logger).Log("msg", "failed to get pci vendor ID", "name", pciDevice.Name(), "err", err)
+			continue
+		}
+		deviceID, err := a.getDeviceID(pciDevice)
+		if err != nil {
+			level.Error(a.logger).Log("msg", "failed to get pci device ID", "name", pciDevice.Name(), "err", err)
+			continue
+		}
+
+		level.Debug(a.logger).Log("msg", "checking pci device", "vendor", vendorID, "device", deviceID)
+
+		cardData, isMonitored := isMonitoredAccelerator(vendorID, deviceID)
+		if !isMonitored {
+			continue
+		}
+		level.Debug(a.logger).Log("msg", "accelerator device found", "vendor", cardData.vendor, "model", cardData.model)
+		devicesCount[cardData] = devicesCount[cardData] + 1
```

Comment:
while counting devices per model/vendor gets you what you want, I wonder if we shouldn't rather expose 1 timeseries per "monitored" card so we can add more labels as we see fit:

````
node_accelerator_card_info{model="V100",vendor="NVIDIA",id="0000:00:02.0",revision="0x00",enable="1"} 1
````

To report metrics to telemetry, we'll need to aggregate anyway:
* with the data model above, we would do `sum by(vendor,model) (node_accelerator_card_info)`.
* otherwise `sum by(vendor,model) (node_accelerator_cards)`.